src/machine/vm.cr

```diff
@@ -1082,6 +1082,11 @@ module VM
         reg = Register.new stack_pop UInt8
         value = reg_read Int32, reg
         STDOUT.puts "#{value}"
+      when Syscall::READ
+        char = STDIN.raw &.read_char
+        if char.is_a?(Char)
+          reg_write Register::R0.byte, char.ord
```

Comment:
> Does this mean that the stack maintains types for its contents at runtime, to distinguish between numbers and register identifiers?

No :) The `push` instruction takes two arguments. The first one is a size specifier which tells the `push` instruction how long the immediate value being pushed onto the stack is. In this case, `t_register` stands for 1 and `counter` stands for `r0`. `counter` gets defined in this file on line 2: `.def counter    r0 ; counter register`.

> I'm curious how my existing implementation is working. 

Your program only works because of a small coincidence. The `sys_puts` syscall actually takes an argument (that you didn't put on the stack before), which tells the machine from which register it should take the value from.

Because the stack is empty at the time this is being executed, it _underflows_ and it reads a byte from the section in memory that is reserved for machine internals (currently all zero). This is documented inside the file [design/execution.md](https://github.com/KCreate/stackvm/blob/master/design/execution.md).

Because that whole section is zero, the `sys_puts` syscall reads a byte from there, and interprets it as a register identifier, in this case for `R0`.

That's the reason this program seems to be working.